smithy-kotlin-codegen/src/main/kotlin/software/amazon/smithy/kotlin/codegen/rendering/ClientConfigProperty.kt

```diff
@@ -310,5 +312,33 @@ object KotlinClientRuntimeConfigProperty {
             """.trimIndent()
             propertyType = ClientConfigPropertyType.Required()
         }
+
+        val tracingClientConfigSymbol = buildSymbol {
+            name = "TracingClientConfig"
+            namespace(KotlinDependency.TRACING_CORE)
+        }
+
+        // TODO support a nice DSL for this so that callers don't have to be aware of `DefaultTracer` if they don't want
+        // to, they can just call `SomeClient { tracer { clientName = "Foo" } }` in the simple case.
+        Tracer = ClientConfigProperty {
+            symbol = RuntimeTypes.Tracing.Core.Tracer
+            baseClass = tracingClientConfigSymbol
+            documentation = """
+                The tracer that is responsible for creating trace spans and wiring them up to a tracing backend (e.g.,
+                a trace probe). By default, this will create a standard tracer that uses the service name for the root
+                trace span and delegates to a kotlin-logging trace probe (i.e.,
+                `DefaultTracer(KotlinLoggingTraceProbe, "<service-name>")`).
+            """.trimIndent()
+            propertyType = ClientConfigPropertyType.Custom { prop, writer ->
+                val serviceName = writer.getContext("service.name")?.toString()
+                    ?: throw CodegenException("The service.name context must be set for client config generation")
+                writer.write(
+                    """override val #1L: Tracer = builder.#1L ?: #2T(#3T, "$serviceName-")""",
```

Comment:
question: Why is this "`$serviceName-"`? (trailing `-`)

Reply:
The argument here is a root prefix—a string which will be prepended to all root-level trace span IDs. If there were no trailing `-` then the service name would blend together with the rest of the span ID (e.g., `DynamoDbPutItem` vs `DynamoDb-PutItem`).

[You suggested in the **aws-sdk-kotlin** PR](https://github.com/awslabs/aws-sdk-kotlin/pull/756#discussion_r1025642640) that the hyphen may be better handled as an implementation detail of the tracer and I agree. I'll modify this to remove the hyphen in the argument and add it in the tracer implementation.

Reply:
Ok I see I missed that on first review